### ISSUE TITLE
The plugin architecture for Umple

### ISSUE THREAD

**Issue by vahdat-ab** [umple/umple#709]
The idea is that to provide a mechanism to have the plugin architecture for Umple. The current idea is on having different code generator based on plugins. The idea first proposed by @bitwizeshift

**Comment by vahdat-ab** [comment]
Prepared and proposed by @bitwizeshift 
**1. Plugin Architecture:**

A pluggable architecture requires various things to work properly, some of which are made easy with newer versions of Java.

First and foremost, a coherent API is necessary for all plugins to extend from. It would be difficult to generically instantiate classes from another Jar without first knowing some details about it. With a coherent API, a plugin loader can be written that instantiates classes using some planned pattern.

Secondly, security is important. It is possible that plugins may attempt to do things that are undocumented and undesirable. To prevent a Plugin from doing things it shouldn't be, it is important to implement Sandboxing and Policies to limit the capabilities of that plugin. Thankfully, this is made very easy as of Java7, since it adds security in their java.security.\* packages. 
This can prevent jars from doing just about anything. This can even limit which files a jar can read, so that they can only work in a small, isolated environment.

Security may not seem so important from the command-line compiler, since it is unlikely to be tampered with, but it could be very important for umpleonline. 

**2. Plugin API**

Most, if not all, Plugin APIs contain a parent interface (or abstract-class) that all plugins descend from. 
This interface normally contains generic information pertaining to the plugin, such as a method for getting the plugin's name, version number, expected permissions, etc.
The API must also contain important hooks to be called by the plugin loader. Example of important hooks are things like "onEnable" or "onDisable" to build-up or tear-down
important plugin information.

The API can also contain important methods necessary for the plugin to satisfy. For example, if the plugin is a Generator Plugin, it could be important for it to provide a "generate(...)" method. 

Extending such features can be handled in various ways, but here are two possible suggestions.

**2.1 Multiple Plugin Types**

The Plugin API would be a system that starts with an UmplePlugin class. This generic superclass contains the basic plugin information necessary to determine
the type of plugin it is, the name, version, etc. The UmplePlugin would take, among possibly other things, an UmpleModel in its construction -- which will be necessary for most 
functionality that is dependent on the model.

All other plugin types would descend from this Plugin, for example GeneratorPlugin, WebPlugin, etc.
All plugins would be required to implement methods "onEnable" and "onDisable", which will be the generic method to build-up and tear-down the plugin.
Specialized plugins like GeneratorPlugin, on the other hand, will have to implement other Generator specific functionalities currently found in the interfaces ILang and CodeGenerator.

An example Umple class of the UmplePlugin would look like this:

```
class UmplePlugin{
  abstract;
  String name;         // The name of this plugin
  String version;      // The version string for this plugin
  Boolean isEnabled;   // Is this plugin enabled?
  File    dataFolder;  // The data folder for this plugin


  onLoad();    // When this plugin loads
  onUnload();  // When this plugin is unloaded
  onEnable();  // When this plugin is enabled
  onDisable(); // When this plugin is disabled

  // other important information ...
}
```

Then any other Plugin types could extend this. Perhaps a Generator one.

```
class GeneratorPlugin{
  abstract
  isA UmplePlugin;
  isA ILang;
  isA CodeGenerator;

  // implement generator-specific methods in here ...
}
```

A plugin need only extend the GeneratorPlugin to create a plugin, and this would stored in a separate jar file.

Then at runtime, all plugins will be instantiated and put into a list or vector, sorted by their plugin type.
At the appropriate time for that form of plugin to be enabled, the code will iterate through all instances of that 
specific plugin. For example, the code for newGenerator() would be changed to something like:

```
for( final GeneratorPlugin plugin : getGeneratorPlugins() ){
  if( plugin.getLang().equalsIgnoreCase(target) ){
    return plugin.getGenerator();
  }
}
// throw exception if one is not found
```

This is one of the possible solutions, the main problem is that it would result in a large collection of plugin classes for different plugins to implement.

**2.2 Listeners and Events**

Another possible method for providing hooks in a more-generic manner would be with the use of an Event system, that fires specific events that get added in plugins.
This would require a whole new interface of Listeners to be added that get called throughout the Umple project at certain times. Whenever a certain event listener is required, Umple would iterate through the list of current listeners and call each one. 

Though this method requires a lot more work, it would also make it a lot more generic, and give one of the easier interfaces to work with since it only be using Listeners to generate new results. This would save having a large hierarchy of type-specific plugins.

It would also break the dependency on UmpleModel on the UmplePlugin side, since the model can then be passed only where needed in event calls.

An example for how such a listener would be implemented in a plugin would be as follows:

```
public class PhpGeneratorPlugin extends UmplePlugin{

  // This would be the listener class
  public class PhpGeneratorListener implement GeneratorListener{

    @Override
    String getGeneratorString(){ ... }

    @Override
    void onGenerate(File[] files, UmpleModel model, /* some other stuff (?) */)

  }

  // ...

  @Override
  void onEnable(){
    // Add the listener to the model
    getLoader().addEventListener( new PhpGeneratorListener() );
  }

  // ...
}
```

This would allow the UmpleModel from the core compiler to register these events, and at the appropriate times will call the registered listeners.

This method is my overall preference, as it would not require various APIs to be written for different plugin-types, and it could allow plugins to be
either small and modular, or large and encompassing (e.g. implementing various listeners in a single plugin).

It's also a scalable solution if there is an inheritance system with a superclass "Listener", and other forms of listeners that extend it being added 
for future installments.

**3. Handling entry-points:**

Knowing that a jar file contains a derivative of UmplePlugin would not be sufficient for instantiating an instance of the plugin, since it doesn't 
have a clear entry-point (it doesn't know the fully-qualified name). There is no 'Main' for a plugin, as there would be for normal software. For this, 
some form of manifest is required to give more details to the plugin loader.

There are two ways that I know of to handle this case.

**3.1 Solution 1:**

Keep a manifest file local to the plugins. This is the approach that Eclipse takes to handling plugins.

All plugins exist in a directory known to the plugin-manager (ex: "plugins"). Inside that directory are a list of directories that contain plugins. Within those 
directories exist manifest.xml files that contain information about the entry-point and necessary dependencies. 
A notable issue with this approach, is that each software isn't distributed easily in just jars, they require individual manifests that exists outside the plugin itself.

**3.1.1 Local Data:**

Because plugins are kept in local directories, local data necessary for the jar can be kept in the same directory the plugin exists in.

**3.2 Solution 2:**

Keep a manifest file inside the plugin jar. This is an approach taken on various servers, including the now-defunct Bukkit and Spigot servers (a custom Minecraft server).

All plugins exist in a directory known to the plugin-manager (ex: "plugins"). Inside that directory are all the jars that correspond to individual plugins. Inside each plugin, at the root of the jar, exists a manifest file (ex: "plugin.xml"). The manifest file contains information about the name of the plugin, the version number, the entry-point, etc.

At runtime the plugin-loader scans the directory for all *.jar files and attempts to read a "plugin.xml" file from it. If it is not found, it skips over it. If it is found, it will attempt to instantiate the specified entry-class from the jar.

**3.2.1 Local Data:**

Since all resources would exist within the jar file, local data would have to be handled somewhere outside the jar. If the plugins are given file-access capabilities, then they can be made to produce a specific plugin directory for that plugin. 

Example: if you have "plugins/MyPlugin.jar", and it needs to have some data written outside the jar. It can produce a directory "plugins/MyPlugin/" which can contain
any further important resources. 

**4. "Dependency Hell"**

This is a topic brought up during the conference call. A system with plugins could potentially be cause for dependency hell.

My only argument against that is that we could prevent it by either completely disallowing dependencies, or allowing dependency APIs to be local to the plugin's personal folder. 
This way each individual jar could have its own dependency version, if necessary. 
For example, lets say MyPlugin.jar is dependent on xyz.jar version 2.0, then the dependency could be located in "plugins/MyPlugin/xyz.jar". This would allow for some other plugin dependent on an older version of xyz.jar (say, version 1.0) to not clash with its required version since they would be in different data folders.

As for possible failures due to future API changes, all I have to say is that as long as the Plugin API is decided and agreed upon early on, there should be few reasons for it to be the cause of failure.

The other topic brought up was to consider compiling everything into Umple.jar instead of individual *.jar files. 
I would personally argue against this since it means that any new plugin produced would require all of Umple to be recompiled for the changes to take effect.

The modularity and drag-and-drop nature of plugins would make usability and distribution of plugins much easier, and would also allow users to be more selective about their choices of plugins. For example, if someone wanted to simply remove support for C++, they could just delete the plugin that adds that capability -- and this wouldn't require a recompilation.
If kept in individual jars, it could help reduce coupling and increase cohesion of modules by giving a narrow interface for plugins to use. 
It could even help cut-back on compilation times, as only the module that has been modified would require recompilation.

About the only downside that I could think of is that the build script would have to be modified to produce each different plugin as they get added; but this could possibly even be restructured such that each plugin becomes their own project (which could potentially help with future maintenance). 

**Side note:**
A lot of this project would be completely orthogonal to the main Umple project, since it would be developing a new API that is currently not in the codebase. As a result, there should be little conflict with current work (at least until hooks are added to the main code)

**Comment by vahdat-ab** [comment]
Comments from @TimLethbridge 
1. The Umple compiler with a plugin architecture would need a major refactoring of its build and test setup. Details of how this would work would need to be outlined before we commit to a plugin setup For example, let's say that php generation was extracted into a plugin. That would mean all of the test architecture that goes with it would also have to be extracted (two of the four levels of testing, including the template tests and the testbed tests), plus all the jet template stuff for php including some refactoring of the UmpleToTemplate hierarchy. All doable, but extremely difficult to get right, especially ensuring that all the other tests work if the php plugin is not installed.
2. Currently using Umple requres no installation. It is 'just a jar'. That is hugely advantageous in that it allows for simple distribution (encouraging use), quick testing and an ability to be certain of the version of Umple that has generated the particular output (which is imprinted in the output files). If plugins are optional (i.e. with the absence of a 'plugins' directory, the compiler still is 'just a jar') and plugins are only used to add new generators or analysis and can't disrupt any of the established mainline features, then isn't so much of an issue. But if the plugin can add syntactic features or new semantics, then the generated output would need to declare that it was generated with a particular combination of plugins. Also, when somebody downloads Umple in such a case they would not be able to 'just use it', they would have to install the plugin directory and any 'necessary' plugins in the right place. Our current philosophy is that we encourage active users to use the latest master jar every time (rebuilt every push to master), and more conservative users to use the 'releases' that are released every few months as a jar. It would be ideal to maintain this.
3. Regarding 'dependency hell'. There are a number of flavors of this that your response didn't consider. A generator plugin would be dependent on a particular version of the core. It is not just API stability that we have to worry about. Generators have to always be consistent with the metamodel elements and generate code from these. If a metamodel element is added, then generators should be updated. If a metamodel element is refactored then generators should be refactored. Currently this is tractable. But if plugins are separately developed, then dependency problems arise. At the very least we would have to have a rule that says that every plugin must be automatically tested whenever changes are made to the core master, an vice-versa, and the plugin developers would have to be responsible for keeping their plugins up to date.
4. Umple is very fast as a compiler. The build takes time, but it is mostly due to timing issues in ant and deliberate injected delays to test time-dependent features of state machines. I would want Umple to continue to be fast, and slowdown at every invocation would have to be measured to ensure it was not significant.
5. Umple is supposed to have an Eclipse plugin. (currently not being built since mid-fall, but that is another issue I am annoyed about). Umple plugins would then become Eclipse plugins too! This would have to be considered.
6. Several of the above issues could be addressed if, at least as an option, we allow two types of plugins: a) Those that are compiled right in with the core; b) the ones found in the plugins directory. That would resolve item 6, 4, 2 and 3.
7. You talk about a 'proof of concept'. I would like it to be something 'real' e.g. a generator of some form of useful output that would be visible in the grammar, testing, UmpleOnline and so on before the end of the semester. I think the idea of a new language or existing langauge would be too much to handle for you. Any ideas for something simpler but real?
8. I prefer your second generic approach with  3.2 Solution 2.
9. I think the main use-cases of this would be to allow people to a) add generators; b) modify generators to suit their needs (e.g. they prefer Java code to look a certain way, so they clone the existing Java generator plugin and make a JavaX plugin that generates java their way); c) add analysis that would spit out different types of warnings, errors and informational messages during parsing; d) add stereotypes that would modify generator behaviour in any generators that cared about that stereotype (which would incidentally add dependencies among plugins).. In the end all generators would become plugin, all the analysis steps would become plugins; and the various stereotypes would become plugins.

**Comment by bitwizeshift** [comment]
I'm just going to summarize all that's been posted here, and what I'm currently working on.

## The Plugin Architecture:
- Plugins are created as separate external, self-enclosed jars
- Umple loads plugins at runtime. Plugins will be stored relative to `umple.jar` in `./plugins/`. 
- Needs a consistent interface to extend from to minimize plugin breakage between versions
- Requires hooks and callbacks throughout Umple to work
- Requires security/sandboxing to prevent rogue plugins (Java7+ provides this)
- Plugins require a manifest to specify the entrypoint. The manifest will be stored at the root of each plugin jar (e.g. in a file `plugin.xml`)
- Needs some way to possibly handle dependency management

## My Current Tasks
- Research general Plugin architecture/design to make an informed suggestion
  - I'm currently creating an outside java plugin project to ensure this idea will work as intended before suggesting it.
  - Ideally I would like to minimize dependencies required for this (i think it can mostly be done with just Java's core libraries)
- Look into how to make generator plugins work with the Eclipse plugin
- Determine if there would be any noticeable performance penalty if using Umple's generators from jars
- Look into ways to keep Umple easily distributed as a single jar
  - This could be handled by bundling core generators/plugins internal to the distributed `umple.jar` as a resource -- and on the first run in copies the resources to the `./plugin/` directory
  - Alternatively, Umple can be released with a set of base languages and features that are not external/accessible, but can be overwritten with other plugins.
- Find possible ways to handle dependencies to avoid "dependency hell" for plugins
- Find a reasonable way to force automated testing for plugins to ensure compatibility with newest versions
- Come up with a real proof-of-concept that is feasible for the time period.
  - I'm tentatively thinking of a logging plugin that prints verbose information during compilation -- and can be modified by a configuration file.

I think that covers just about everything discussed above. Most of it is just a lot of research that's required first to make an informed suggestion before continuing.

**Comment by Nava2** [comment]
I think @aforward should be involved in this discussion. He and I have talked about this extensively in the past. One of the big things that this also leads into is separating the monolithic repository into their own repositories. I haven't spent much time figuring this out, @aforward had some thoughts on that one. He is very busy though, so I'm unsure if he'll be able to help. 

In regards to the "dependency hell," this should be straight forward using Ivy. I found a programmatic API [here](https://cwiki.apache.org/confluence/display/IVY/Programmatic+use+of+Ivy). This would likely be the best approach for dealing with dependencies. Include an `ivy.xml` within the plugin and load it via Java's resource loader and you can call the API to pull in the dependencies. 

I am _really_ hesitant to package the plugin jars _inside_ the `umple.jar`, but having them automatically fetched would make sense. For example, if using the compiler and passing `php`, it would check if the plugin exists. If it does not exist, it would download and "install" it. Then, just use the plugin after that using the sandbox and security measures. 

This is just my thoughts on this one, I like the idea, but it's a pretty **massive** undertaking, careful.

**Comment by bitwizeshift** [comment]
@Nava2:
Thank you for your input.

I wasn't aware Ivy actually had a programmatic API like that, that will definitely prove useful -- thanks for the link!
One of the biggest problems with the dependencies that @TimLethbridge is concerned about is that plugins may become out of sync with the core Umple compiler. As changes are made to the metamodel, it could be possible to break some of those plugins -- and so it would be necessary to attempt to keep these synchronized and still passing all required tests.

> I am really hesitant to package the plugin jars inside the umple.jar, but having them automatically fetched would make sense.

That's definitely a better idea. The only reason I suggested packing plugins inside the jar was to try to keep the current distribution capabilities of it being 'just a jar', since it's easier to download and run.

Though I'm not sure if downloading & installing it _only_ when passed a generator string that it doesn't have would be the best way to go about it. I feel like it would be an unexpected action for the compiler to suddenly need internet access for a module, when the command issued was to generate a language.

That being said, we could make it download _all_ the "standard" plugins on first use -- perhaps through the use of a special command. Or, alternatively, have the Umple Core contain a few standard hard-coded features that can instead be _overridden_ by plugins. This would guarantee some features exist from the beginning, but also allow it to be customizable.

Things like the Java generator or graphical generators would be need to be default, as they are necessary for the core compiler and Umple Online. So something like that could be hard-coded into the core -- but a plugin can override the basic feature-set normally provided in it. 

That's my current thought on the matter though, but I do welcome input

**Comment by aforward** [comment]
The simplicity of having _just a jar_ is a huge benefit.

Before moving away from this, I would first consider trying to separate the code base into smaller projects (which is already is, just not in separate repositories); but having the build mechanism pull things together (which is already does, again just not in separate repos).

The ability to dynamically load code sounds awesome, but I don't think the use case outweighs the simplicity of what's currently there.

Please don't make umple.jar be dependent on other jars.

**Comment by TimLethbridge** [comment]
OK, so Andrew is reflecting my concerns. I think simplicity needs to be kept forefront. A lot of software systems have moved in the direction of 'flexibility', 'pluggability' etc. and have just become so complex. Even Umple loading dependencies now is more complex to build than it was, and more at risk caused by changes to those dependencies

 I think we can separate the notion of having the ability to plug in from needing to have separate jars. I say we stick with one jar for everything 'normal', but that the plugin capability should allow a possibly zero set of add-ons or overrides. Most users would continue to just use a separate Jar, but if they put an umplepluginsd directory as a sibling, then Umple would look there to find special features.

The concept of having the default capabilities use the plugin architecture should not be so quickly discounted as it would allow us to test that architecture, even if the plugins reside in the same jar. This goes against some philosophies of plugins, but is consistent with our overriding philosophy of simplicity.

That said, I think first steps need to be to do a certain amount of refactoring. The analysis code is 'half-refactored' into separate classes/files for example. This needs to be completed.

**Comment by vahdat-ab** [comment]
One of my main concerns about this architecture is about using metamodel of Umple. We continuously improve Umple metamodel and this dynamic nature will kill all plugins. Plugin developers need to have a full metamodel in order to be able to build their plugins (for example, a code generator). It's really hard and tedious work for plugin developers to keep updating their work with our progress.

I believe that the work should be done first on parts of Umple which have more stable nature. This makes more sense to me. For example, why we don't let people make your own analysis phase through plugins? This is much more stable than letting people two work on our metamodel which is dynamic? another example, why we don't let people write their own checkers?

I also think that we need to focus more right now on the current architecture of Umple and refactor it to more independent layers. Plugin architecture may not be successful with current state of Umple architecture because it's been developed in a agile way and its modular and layered architecture have changed a lot. 

If we start developing plugin architecture we end up with many discussions in which we would say "oh" this cannot be done because we have so crosscutting nature in Umple architecture, in particular, analysis and checking parts.

Therefore, designing a great layered architecture for Umple with the goal of reuse will help a lot the future plugin architecture. By saying reuse, it doesn't mean someone can easily use our modular layers, in fact we can also easily bring other works to our architecture.

Finally, I believe improving layered architecture and refactoring Umple is worth because it will open doors for many collaboration, rearch- and technology-wise. Finally, It will provide the base for plugin architecture.

**Comment by bitwizeshift** [comment]
@aforward 
I'm not suggesting that the entire project become a skeleton designed around plugins or anything; I'm mostly suggesting the possibility of writing ways to override original features or add new ones through an external interface. 

This way it can still have the benefit of a single monolithic jar for most users, but also allow powerful additions for those that wish to extend it without having to clone or fork the umple project to add it.

> Before moving away from this, I would first consider trying to separate the code base into smaller projects (which is already is, just not in separate repositories); but having the build mechanism pull things together (which is already does, again just not in separate repos).

I definitely agree that the different modules is a good idea that should at some point be done. It could possibly be handled straight in git, using the git submodule capability -- or as just a step in the ant build.

That being said, I think that may be deviating a lot further from my initial project scope.

@TimLethbridge 

>  I say we stick with one jar for everything 'normal', but that the plugin capability should allow a possibly zero set of add-ons or overrides.

This was one of the things I was suggesting, this way the current state of Umple doesn't have to change, and the distribution is kept as a single jar. But in the event that someone wanted to add a new feature, it can be done so through plugins.

> That said, I think first steps need to be to do a certain amount of refactoring. The analysis code is 'half-refactored' into separate classes/files for example.

Are you referring to the `*Analyzer` classes themselves? Looking through the javadocs, it appears that there are 60-or-so classes that are residing in around 3 ump files. 
What kind of refactoring are you thinking of? I haven't looked in-detail into the Analyzer yet, and am not sure if it should be broken into further, smaller classes -- or if the data should be better reorganized into different files. Umple has a different philosophy than Java's "1-class per file" nature, and so I don't know if the current file organization is normal or not.

I will make sure to start with cleanup though; a lot of the Umple code would need to be prepped for the addition of plugin support.

@vahdat-ab 

> why we don't let people make your own analysis phase through plugins?

My original suggestion for this project didn't exclude that type of plugin; I just focused on Generators as an example -- partially because I don't think I fully realized the amount of issues that I'd be running into. 

My main idea is for a generic plugin system that could easily be extended by anyone in the future so that plugins could add even more functionality. This _could_ mean that, for the project, it only supports simpler hooks like the analyzer or logging, or other command-line functionality -- but in the future it could be extended to add other support too. This could possibly mean that Generators won't be able to be a plugin in the beginning as a result, but can be added once the metamodel becomes more stable.

> I also think that we need to focus more right now on the current architecture of Umple and refactor it to more independent layers.

I definitely agree that a refactor would be largely beneficial. That said, do you have any suggestion for where to even begin in such a case? Are you thinking the Analyzer as well? I've been through a lot of Umple's source code, but I'm not hugely sure on what are in the different layers short of analysis, parsing, and code generators.

**Comment by bitwizeshift** [comment]
Sorry, accidentally misclicked the close button.

**Comment by bitwizeshift** [comment]
Just a follow-up from Wednesday's call

For this project, in order to lay the framework for a proper plugin architecture, I will be doing the following:
- Removing coupling and extracting the Parser (`RuleBasedParser`) and all subclasses into an individual project/Module
- Use Git SubModule for the Parser class, so that the module is embedded in the project.
  - This is so that Ant does not require repeat downloads to compile the project  
- Extract the Analyzer into different classes, so it isn't entirely in `UmpleInternalParser`. This has partially been done, so I just need to finish it, implement it, and test it (probably a lot easier said than done).
- Modify Ant build to compile removed modules into the full version of Umple
- General code refactoring/clean up. Possibly removing dependencies.
- If possible, where applicable, I'd like to document as well so that the javadocs don't have empty descriptions
- If time: Extract the Code generator (the umple template java generator) into a separate project, as it has further applications outside Umple as well.

Rationale for extracting the module into individual Git repos:
- The module will contain an individual commit log, which will be easier to see changes to the repo
- Keeping individual repos for generic classes encourages code reuse, and can help other potential authors/implementors (particularly in the case of `RuleBasedParser`)
- Git is able to extract changes to specified files into a new git repo, so that this new repository will still keep it's original change history -- but only for the important files (I believe the command is `git filter-branch`)

One question I have though, specifically with regard to `RuleBasedParser`: If this is going to be extracted into a separate project for possible reuse, would it be better to rewrite the API strictly in Java so it is more accessible (i.e. de-umplify the Umple code)? This would make it easier for other groups to fork/modify/use without any knowledge of Umple.

**Comment by vahdat-ab** [comment]
Umple completely is written by Umple and we get lots of benefits from it. Keeping Umple parser in Umple is main advantage (and we must keep it written in Umple as well). Image how easily we can implement it in c++ because of our generator.
As Umple grows, we will be able to add more modeling elements to Umple itself and finally we'll e able to generate different target langauges with minimum manual code.

**Comment by aforward** [comment]
As Vahdat stated, the parser should remain in umple.

I strongly  discourage sub modules (and subtrees).  They are difficult to use correctly and even worse if you are not a gitpert.

Final note, the work described below should be labelled more under a vanilla refactoring, as it does not pave way for a plugin architecture, but rather we are extracting a parser component that others could use (instead of enabling others to plug new stuff into umple).  

This work is still important, but I wouldn't label it as a plugin architecture to avoid confusion.

## 

aforward

**Comment by bitwizeshift** [comment]
@vahdat-ab 

Alright thanks, I was just checking. I wasn't sure if it should be changed for accessibility reasons (i.e. if someone wished to implement the project in Java, this way they wouldn't have to use Umple in order to do it)

@aforward 

In that case I will look further into it being pulled in from another project during the build process. Would you happen to know if it's possible to use ant to only draw in sources if they have changed, and to cache a local copy of it? Tim's main apprehension with it being pulled in during the build was that it would require frequent downloading of the modules during the build process which would increase build times.

If I could somehow cache the result and perhaps compare git commit hashes, then the build process would only be longer when it has to update changes -- but I'd like a way that doesn't require the entire repo to be pulled just for a build.

> the work described below should be labelled more under a vanilla refactoring, as it does not pave way for a plugin architecture

That's actually a good point. Should this be opened as a separate issue in that case?

**Comment by vahdat-ab** [comment]
We might need to have a look at this https://codingkilledthecat.wordpress.com/2012/04/28/why-your-company-shouldnt-use-git-submodules/
It talks about the problems with the submodules and candidate solutions. For me, those solutions are not good again for our purposes.
@bitwizeshift Yes. You can create another issue with the goal that currently we deal with. Please add all people who participated in this issue and connect the issue to this issue as well.
